Add architecture support for mips(64 bit mainframe)

sunny868 opened the pull request:
This version of PR is based on #894.
The previous fork is too old, resulting in many conflicts, so I redo the fork and have this new PR.

adrianreber commented:
There are s few test failures which do not seem to be flakes. I have not seen any other pull request showing those errors. So something in your pull requests seems to be breaking:

 * zdtm/static/sigpending(h)
 * zdtm/static/pthread01(h)

adrianreber commented on .travis.yml:
```diff
@@ -86,6 +86,10 @@ matrix:
       env: TR_ARCH=armv7-cross
       dist: bionic
+    - os: linux
+      arch: mips64
```
This does not work. Travis does not provide `mips64`. Travis should just fail but this test is running on `x86_64` if you look at the log files. You need to look at how we used to run docker with qemu emulation for aarch64, ppc64le and s390x before Travis had native support for those architectures.

adrianreber replied:
Or you could try to install a cross compiler into the Travis environment and provide cross-compilation test.

sunny868 replied:
I am sorry. I am not understand,can you tall me how to try to install a cross compiler into the Travis environment? or I should add a file like Dockfile.mips.hdr ?

adrianreber replied:
Please have a look how it is done for arm here https://github.com/checkpoint-restore/criu/blob/criu-dev/scripts/build/Dockerfile.armv7-cross . This uses a cross compiler.

The other way is to use docker and qemu to emulate another architecture:
https://github.com/checkpoint-restore/criu/blob/v3.13/scripts/build/Dockerfile.aarch64.hdr
https://github.com/checkpoint-restore/criu/blob/v3.13/scripts/build/binfmt_misc

The whole binfmt, qemu, docker combination has been removed from the criu-dev branch as all other CRIU architectures (arm, aarch64, s390x, ppc64le, x86_64) are natively supported by Travis. You have to decide what is better for mips: cross-compiler or the qemu docker combination.

sunny868 replied:
I have add scripts/build/Dockerfile.mips64el.hdr. but I don't know how to make it work

adrianreber commented on criu/arch/mips/include/asm/restorer.h:
```diff
@@ -0,0 +1,60 @@
+		     "move $8, %3				\n"	\
+		     "move $9, %4				\n"	\
+		     "break  					\n"	\
+		     : "=r"(ret)					\
```
Having just worked on this for all other supported architectures, this looks wrong. You need to call the clone() syscall to create a new process. This just seems to move registers around without doing any syscall.

sunny868 replied:
thanks ,I will support multi-thread for mips architecture,then I will check it.

sunny868 replied:
I have support multi-thread for mips architecture

adrianreber commented on criu/arch/mips/include/asm/restorer.h:
```diff
@@ -0,0 +1,60 @@
+#define RUN_CLONE3_RESTORE_FN(ret, clone_args, size, args, \
+			      clone_restore_fn)	do { \
+	pr_err("This architecture does not support clone3() with set_tid, yet!\n"); \
+	ret = -1; \
```
And doing it for clone3() would also be nice. I was to test the clone3() code on all other architectures natively, but I have no access to mips64.

sunny868 replied:
I am working on kernel version v4.19 for mips architecture.  so I can`t check clone3().

adrianreber commented on images/core.proto:
```diff
@@ -84,7 +86,7 @@ message thread_core_entry {
 	optional uint32			sched_policy	= 4;
 	optional uint32			sched_prio	= 5;
-	optional uint64			blk_sigset	= 6;
+	repeated uint64			blk_sigset	= 6;
```
Not sure if this has been discussed before. But does this break our protobuf files if we redefine an object from optional to repeated?

0x7f454c46 replied:
I think, changing optional to repeated instead of adding a new filed - that what I've asked in the previous review. Quoting from language spec:

> `optional` is compatible with `repeated`. Given serialized data of a repeated field as input, clients that expect this field to be optional will take the last input value if it's a primitive type field or merge all input elements if it's a message type field. 

https://developers.google.com/protocol-buffers/docs/proto

adrianreber replied:
Thanks. Sounds good.

rppt commented on compel/src/lib/ptrace.c:
```diff
@@ -110,18 +110,26 @@ int ptrace_swap_area(pid_t pid, void *dst, void *src, long bytes)
 int __attribute__((weak)) ptrace_get_regs(int pid, user_regs_struct_t *regs) {
+#ifdef CONFIG_MIPS
+     	int ret = ptrace(PTRACE_GETREGS, pid, NULL, regs);
+#else
```
If `ptrace(PTRACE_GETREGSET)` does not work for MIPS for some reason, you would need to override the entire `ptrace_{get,set}_regs` in `compel/arch/mips/src/lib/infect.c`. See for example as s390 and x86 do it.

rppt commented on criu/parasite-syscall.c:
```diff
@@ -179,8 +179,11 @@ int parasite_dump_thread_seized(struct parasite_thread_ctl *tctl,
-
+#ifdef CONFIG_MIPS
+	memcpy(tc->blk_sigset, compel_thread_sigmask(tctl), sizeof(k_rtsigset_t));
+#else
```
Any reason why MIPS should be special?

sunny868 replied:
sizeof(k_rtsigset_t) is 16byte (two uint64_t) in MIPS platform. I set n_blk_sigset = 2 in pstree.c

rppt commented on criu/pstree.c:
```diff
@@ -113,6 +113,11 @@ CoreEntry *core_entry_alloc(int th, int tsk)
+#define CR_BLK_SIGSET_SIZE 2
+			core->thread_core->n_blk_sigset = CR_BLK_SIGSET_SIZE;
+			core->thread_core->blk_sigset = xptr_pull_s(&m, CR_BLK_SIGSET_SIZE * sizeof(core->thread_core->blk_sigset[0]));
+#endif
```
Ditto

0x7f454c46 commented on compel/arch/mips/plugins/std/syscalls/syscall_64.tbl:
```diff
@@ -0,0 +1,115 @@
+##TODO for kernel
+__NR_fsopen			5400		sys_fsopen		(char *fsname, unsigned int flags)
+__NR_fsconfig			5411		sys_fsconfig		(int fd, unsigned int cmd, const char *key, const char *value, int aux)
+__NR_fsmount			5422		sys_fsmount		(int fd, unsigned int flags, unsigned int attr_flags)
```
https://github.com/checkpoint-restore/criu/pull/894#discussion_r367548707
hmm?

sunny868 replied:
yes,I did make a mistake

0x7f454c46 commented on compel/arch/mips/scripts/compel-pack-compat.lds.S:
```diff
@@ -0,0 +1,4 @@
+OUTPUT_ARCH(mips)
+EXTERN(__export_parasite_head_start)
+
+/*TODU*/
```
https://github.com/checkpoint-restore/criu/pull/894#discussion_r367553003

0x7f454c46 commented on compel/arch/mips/src/lib/include/handle-elf.h:
```diff
@@ -0,0 +1,13 @@
+
+#define ELF_MIPS
+#define __handle_elf				handle_elf_mips
+//#define handle_elf_mips __handle_elf
```
https://github.com/checkpoint-restore/criu/pull/894#discussion_r367554972

0x7f454c46 commented on compel/arch/mips/src/lib/include/ldsodefs.h:
```diff
@@ -0,0 +1,151 @@
+
+#define DL_RO_DYN_SECTION 1
+
+//#include_next <ldsodefs.h>
```
https://github.com/checkpoint-restore/criu/pull/894#discussion_r367555517

0x7f454c46 commented on compel/arch/mips/src/lib/include/uapi/asm/cpu.h:
```diff
@@ -0,0 +1,14 @@
+ */
+
+#define NCAPINTS			(12) /* N 32-bit words worth of info */
+#define NCAPINTS_BITS			(NCAPINTS * 32)
```
https://github.com/checkpoint-restore/criu/pull/894#discussion_r367557672

0x7f454c46 commented:
Please, see how-to add a build test for mips64 to Travis in https://github.com/checkpoint-restore/criu/pull/979/commits/894bc47733089fea6bdf899ade75d7c61916c7a2
You should just change `ppc64el` and `powerpc64le-linux-gnu` to something appropriate like `mips64le`. To verify you can run `docker build` locally or push it into your branch and see the results on Travis-CI.

sunny868 commented:
> Please, see how-to add a build test for mips64 to Travis in [894bc47](https://github.com/checkpoint-restore/criu/commit/894bc47733089fea6bdf899ade75d7c61916c7a2)
> You should just change `ppc64el` and `powerpc64le-linux-gnu` to something appropriate like `mips64le`. To verify you can run `docker build` locally or push it into your branch and see the results on Travis-CI.

cross-compiler is not suitable for mips ,where many packages are missing. I have push a scripts/build/Dockerfile.mips64el.hdr for using qemu docker combination .

sunny868 commented:
> Please, see how-to add a build test for mips64 to Travis in [894bc47](https://github.com/checkpoint-restore/criu/commit/894bc47733089fea6bdf899ade75d7c61916c7a2)
> You should just change `ppc64el` and `powerpc64le-linux-gnu` to something appropriate like `mips64le`. To verify you can run `docker build` locally or push it into your branch and see the results on Travis-CI.

mips64el-cross build base docker has succeeded. please review my code ,thank you very much.

sunny868 commented:
@0x7f454c46 results of run zdtm is https://gist.github.com/sunny868/4fc6c12b8f07592a0b5918a2d9208584 

there are some error , I need to check it ?

adrianreber commented:
@sunny868 Your changes broke some tests on other architectures. All existing test cases in Travis should still be successful. Have a look at the Travis results and make sure all the tests are still green.

0x7f454c46 commented:
@sunny868 please, run tests on mips with `--keep-going` option - this way we can see overall situation. You did great with mips64 cross-compilation test - now when this merges it'll be possible to verify that no changes breaks mips support.
As @adrianreber said, please check the results on Travis-CI: it seems that pthread and sigpending tests are broken on x86.
Also, when you finish, please rebase it over criu-dev and force-push it, squashing fixes/fixups into  commits those introduce them. It will be much easier for reviewers to see the end version.

sunny868 commented:
zdtm test result for mips64el in https://gist.github.com/sunny868/5c50780f785fff176d057945d68f4504

but there are a problem :test is not finish normally.It is pause in "29907 ?        Ssl    0:04 /usr/libexec/packagekitd" ,can not go on.can you give me a help?

sunny868 commented:
> @sunny868 Your changes broke some tests on other architectures. All existing test cases in Travis should still be successful. Have a look at the Travis results and make sure all the tests are still green.

I have a changes for this ,commit a37e94256f191ea981f0de14cac979a883ab8f06  but why The Travis CI have not working to build ?

sunny868 commented:
> Also, when you finish, please rebase it over criu-dev and force-push it, squashing fixes/fixups into commits those introduce them. It will be much easier for reviewers to see the end version.

Sorry,I can't quite seize your meaning.can you give me a example ? thank you very much

sunny868 commented:
> > Also, when you finish, please rebase it over criu-dev and force-push it, squashing fixes/fixups into commits those introduce them. It will be much easier for reviewers to see the end version.
> 
> Sorry,I can't quite seize your meaning.can you give me a example ? thank you very much

maybe also like  s390 https://patchwork.criu.org/series/1716/. ?

adrianreber commented:
Travis is a bit slow the last days. The tests are actually running in Travis, but the result is never reported back to github.

You should see your results here: https://travis-ci.org/github/checkpoint-restore/criu/pull_requests

adrianreber commented:
> Sorry,I can't quite seize your meaning.can you give me a example ? thank you very much

You should rebase your work always on top of the latest criu-dev branch. Looking at the commits in this pull request I see merge commits. There should be no merge commit. Just a number of commits on top of criu-dev. Not sure how you were using git, but instead of a `git pull` to get the latest criu-dev changes, probably a `git pull --rebase` would be better.

You also need to group your changes logically together.

Also, do not fix your commits with new commits pushed to branch. Rework your commits to be correct and contain the changes from the reviews. One example where I have been involved was that you first provided a broken CLONE assembler wrapper for threaded processes and then in a later commit you fixed it. There is no need, for a pull request to contain fix up commits, just change your commits to be correct the first time.

0x7f454c46 commented:
Hi, @sunny868,
I see you've rebased over criu-dev and merged fixups - that looks good.
Passing Travis-CI build seems promising to me: https://travis-ci.org/github/checkpoint-restore/criu/builds/668576184
Please, finish addressing comments I've left for the previous PR here: https://github.com/checkpoint-restore/criu/pull/894
I still see many of them not addressed, i.e: https://github.com/checkpoint-restore/criu/pull/894#discussion_r367577296
Thanks,
Dmitry

sunny868 commented:
> Hi, @sunny868,
> I see you've rebased over criu-dev and merged fixups - that looks good.
> Passing Travis-CI build seems promising to me: https://travis-ci.org/github/checkpoint-restore/criu/builds/668576184
> Please, finish addressing comments I've left for the previous PR here: #894
> I still see many of them not addressed, i.e: [#894 (comment)](https://github.com/checkpoint-restore/criu/pull/894#discussion_r367577296)
> Thanks,
> Dmitry

Thanks for the reminder. I have fix it again for the PR #894

0x7f454c46 commented on include/common/arch/mips/asm/non-atomic.h:
```diff
@@ -0,0 +1,109 @@
+#ifndef _ASM_GENERIC_BITOPS_NON_ATOMIC_H_
+#define _ASM_GENERIC_BITOPS_NON_ATOMIC_H_
+
+#include <asm/types.h>
+#include "common/arch/mips/asm/bitops.h"
+#include "common/arch/mips/asm/utils.h"
```
I don't see this "non-atomic.h" header used anywhere.
Can it be dropped?

sunny868 replied:
Yes, It's really useless，I have dropped it.

0x7f454c46 commented on compel/arch/mips/src/lib/infect.c:
```diff
@@ -0,0 +1,268 @@
+		unsigned long arg5,
+		unsigned long arg6)
+{
+
+/*refer to glibc-2.20/sysdeps/unix/sysv/linux/mips/mips64/syscall.S*/
+
+	user_regs_struct_t regs = ctl->orig.regs;
+	int err;
+
+	regs.regs[2] = (unsigned long)nr; //syscall_number will be in v0
+
+	regs.regs[4] = arg1;
+	regs.regs[5] = arg2;
+	regs.regs[6] = arg3;
+	regs.regs[7] = arg4;
+	regs.regs[8] = arg5;
+	regs.regs[9] = arg6;
+
+	err = compel_execute_syscall(ctl, &regs, code_syscall);
+	*ret = regs.regs[2];
+
+	 return err;
+	
+}
```
you have excessive empty lines at the function begin and end

0x7f454c46 commented on compel/arch/mips/src/lib/infect.c:
```diff
@@ -0,0 +1,268 @@
+	}
+
+
+	  /*Restart the system call*/
+	if(regs->regs[0]){
```
Align the comment with tabs, put spaces between `if` and brackets

0x7f454c46 commented on compel/Makefile:
```diff
@@ -17,6 +17,10 @@ ldflags-y		+= -r
 # UAPI inclusion, referred as <compel/...>
 ccflags-y		+= -I compel/include/uapi
 
+ifeq ($(ARCH),mips)
+ccflags-y		+= -I compel
+endif
+
```
Let's have this `-I compel` for all platforms, rather than having excessive `ifeq`

sunny868 replied:
ok

sunny868 replied:
ok,I have do it

0x7f454c46 commented on criu/arch/mips/vdso-pie.c:
```diff
@@ -0,0 +1,22 @@
+			struct vdso_symtable *sto, struct vdso_symtable *sfrom,
+			bool compat_vdso)
+{
+    pr_warn("NO implement on mips\n");
+    return 0;
```
```
   pr_err("Vdso proxification isn't implemented on mips\n");
   return -1;
```
It's a critical error - it happens only when vdso is different from the one on dump.
Which usually a different kernel.

0x7f454c46 commented:
@sunny868 also, there is some whitespace garbage:
- trailing whitespaces
- excessive empty lines
- align with spaces
If you could clean those up a bit too, it'd be good.

0x7f454c46 commented:
@avagin, @cyrillos, @rst0git, @rppt, @adrianreber, guys, please spend some time on this.
I think it is very close to be merge-ready.
Travis-ci tests pass, a new cross-build for mips is present.
I have very few comments and mostly trivial to fix.
There are some zdtm tests not passing on mips, but I think a decent amount passes and those left @sunny868 can fix after the merge.
Thanks!

adrianreber commented:
I cannot see Travis right now, but overall this looks pretty good, indeed. I cannot say anything about the actual MIPS code, but the structure of the commits and the content of each commits looks like it should be. First the code is added and then enabled. No existing test cases are breaking and there is cross-compilation target for MIPS. Just like it should be.

One thing which is also still missing is to make sure to not enable clone3() on MIPS as the necessary assembler code is missing. There used to be an ifarch (or something like this) before I added the code for all architectures. This is an old version of kerndat.c which was active when only X86_64 and S390X clone3() support was enabled:

https://github.com/checkpoint-restore/criu/blob/0f0564a13a0f5b02809c9d5c782d3c2b5c1eef7c/criu/kerndat.c#L995

@sunny868 you need something like this for MIPS to never use clone3() as long as the assembler wrapper is not implemented.

sunny868 commented:
> @sunny868 you need something like this for MIPS to never use clone3() as long as the assembler wrapper is not implemented.
Now MIPS code just run on kernel v4.19 which no clone3, so I can't verify its correctness even if I impliment clone3(). but when I run criu on kernel v4.19now,pid will return -1 in criu/kerndat.c,

	pid = syscall(__NR_clone3, &args, sizeof(args));

	if (pid == -1 && (errno == ENOSYS || errno == E2BIG)) {
		kdat.has_clone3_set_tid = false;
		return 0;
	}
so has_clone3_set_tid also be false,then do I must add #if !defined(CONFIG_MIPS) for it ?

adrianreber commented:
> > @sunny868 you need something like this for MIPS to never use clone3() as long as the assembler wrapper is not implemented.
>  Now MIPS code just run on kernel v4.19 which no clone3, so I can't verify its correctness even if I impliment clone3(). but when I run criu on kernel v4.19now,pid will return -1 in criu/kerndat.c,

But if you run CRIU MIPS on kernel 5.6 CRIU would detect clone3() support and enable it and fail during restore because the assembler wrapper is missing.

> 
> ```
> pid = syscall(__NR_clone3, &args, sizeof(args));
> 
> if (pid == -1 && (errno == ENOSYS || errno == E2BIG)) {
> 	kdat.has_clone3_set_tid = false;
> 	return 0;
> }
> ```
> 
> so has_clone3_set_tid also be false,then do I must add #if !defined(CONFIG_MIPS) for it ?

Something like this should work for you:
```
#if defined(CONFIG_MIPS)
	/*
	 * Currently the CRIU PIE assembler clone3() wrapper is
	 * not implemented for MIPS.
	 */
	kdat.has_clone3_set_tid = false;
	return 0;
#endif
```

rppt commented on include/common/arch/mips/asm/bitsperlong.h:
```diff
@@ -0,0 +1,10 @@
+#ifdef CONFIG_MIPS
+# define BITS_PER_LONG 64
+#else
+# define BITS_PER_LONG 32
```
Are we going to support mips32? Even if yes, the 32-bit version of `BITS_PER_LONG` can be added then.

rppt commented on compel/src/lib/infect.c:
```diff
@@ -837,6 +837,45 @@ void compel_relocs_apply(void *mem, void *vbase, size_t size, compel_reloc_t *el
 			int *where = (mem + elf_relocs[i].offset);
 			*where = elf_relocs[i].value + elf_relocs[i].addend + (unsigned long)vbase;
+#ifdef CONFIG_MIPS
+                /*
```
It would be nice to have all the MIPS-specific relocs handling in a separate function.

rppt commented on compel/src/lib/handle-elf.c:
```diff
@@ -16,8 +16,9 @@
-piegen_opt_t opts = {};
-
+#ifdef CONFIG_MIPS
+#include "arch/mips/src/lib/include/ldsodefs.h"
```
This could be simply
`#include "ldsodefs.h" and than you can omit addition `-Icompel` to the `Makefile`:
```
diff --git a/compel/Makefile b/compel/Makefile
index d745d0eb7..de9318c42 100644
--- a/compel/Makefile
+++ b/compel/Makefile
@@ -16,7 +16,6 @@ ldflags-y		+= -r
 #
 # UAPI inclusion, referred as <compel/...>
 ccflags-y		+= -I compel/include/uapi
-ccflags-y		+= -I compel
 
 lib-name		:= $(LIBCOMPEL_A)
 lib-y			+= src/lib/log.o
diff --git a/compel/src/lib/handle-elf.c b/compel/src/lib/handle-elf.c
index ded1908f6..e97d75026 100644
--- a/compel/src/lib/handle-elf.c
+++ b/compel/src/lib/handle-elf.c
@@ -17,7 +17,7 @@
 #include "log.h"
 
 #ifdef CONFIG_MIPS
-#include "arch/mips/src/lib/include/ldsodefs.h"
+#include "ldsodefs.h"
 #endif
 /* Check if pointer is out-of-bound */
 static bool __ptr_oob(const uintptr_t ptr, const uintptr_t start, const size_t size)
```

avagin commented:
@sunny868 Great job!!!
@0x7f454c46 @adrianreber @rppt Thank you for reviewing and helping with this PR.

sunny868 commented:
@adrianreber @0x7f454c46 @avagin @rppt Thanks for your reviewing and help me.

0x7f454c46 commented:
@sunny868, good work!
Please, keep working on the mips criu port :-)
Check zdtm failures on mips and try to address them.
Thanks!

sunny868 commented:
> @sunny868, good work!
> Please, keep working on the mips criu port :-)
> Check zdtm failures on mips and try to address them.
> Thanks!

ok
